### PR TITLE
Support array remove with negative index

### PR DESCRIPTION
**Pull request by ashanbrown**
Also return an error when when a negative index is out of range

### CODE REVIEW

File: patch.go [hunk 1]
```diff
@@ -389,17 +389,13 @@ func (d *partialArray) add(key string, val *lazyNode) error {
...
 
-	if idx < 0 {
-		idx *= -1
-
-		if idx > len(ary) {
-			return fmt.Errorf("Unable to access invalid index: %d", idx)
-		}
-		idx = len(ary) - idx
-	}
-	if idx < 0 || idx >= len(ary) || idx > len(cur) {
```
Comment (ashanbrown) [inline patch.go]: I believe the > and >= checks here are redundant so I removed one of them.

Comment (ashanbrown) [reply patch.go]: Also, multiplying the index by -1 before reporting an error made the error message a bit misleading, so I check the index before altering it.

**Comment by ashanbrown** [comment]
This was a big fat fail.  Gonna have to investigate the differences between the patch libraries some more.  For one, the original repo doesn't seem to support overwriting a null value.

**Comment by ashanbrown** [comment]
Sorry about that last comment.  Put it on the wrong PR.  ;) . I think this PR should work fine.  I'm resolving the changes after you merged my other PR (thanks for that!).

**Comment by evanphx** [comment]
I don’t believe negative indexes are in the RFC...

**Comment by clagraff** [comment]
I did a quick check @evanphx 

[JSON Patch RFC 6902 section 4.1](https://tools.ietf.org/html/rfc6902#section-4.1):
> If the "-" character is used to index the end of the array (see [RFC6901]), this has the effect of appending the value to the array.

[JSON Pointer RFC 6901 Section 4](https://tools.ietf.org/html/rfc6901#section-4)
> Note that the use of the "-" character to index an array will always result in such an error condition because by definition it refers to a nonexistent array element.

So it seems like only `-` is explicitly supported by the RFC. No other negative indices like`-1`, `-2`, `-3`, etc etc.

---

The repo currently _supports_ the use of negative indexes, as indicated in [this test](https://github.com/evanphx/json-patch/blob/master/patch_test.go#L71).

So I'd think this PR would be a good addition for now, and if you wanted to remove the ability to specify negative indices that might be best tackled in a separate pr?

**Comment by ashanbrown** [comment]
FWIW, I created this PR just because I  wanted the error handling.  But when I saw that “add” handled negative indices, I figured we needed to support that as well.  We personally don’t use or expect them.

**Pull request merged**